tempesta_fw/apm.c

```diff
@@ -871,59 +865,140 @@ tfw_apm_stats(void *apmdata, TfwPrcntlStats *pstats)
  * All APM Stats users must use the same set of percentiles.
  */
 int
-tfw_apm_prcntl_verify(TfwPrcntl *prcntl, unsigned int prcntlsz)
+tfw_apm_pstats_verify(TfwPrcntlStats *pstats)
 {
 	int i;
 
-	if (prcntlsz != ARRAY_SIZE(tfw_apm_prcntl))
+	if (pstats->psz != ARRAY_SIZE(tfw_pstats_ith))
 		return 1;
-	for (i = 0; i < prcntlsz; ++i)
-		if (prcntl[i].ith != tfw_apm_prcntl[i].ith)
+	for (i = 0; i < pstats->psz; ++i)
+		if (pstats->ith[i] != tfw_pstats_ith[i])
 			return 1;
 	return 0;
 }
 
 static inline void
+tfw_apm_data_put(TfwApmData *data)
+{
+	if (atomic_dec_and_test(&data->refcnt))
+		kfree(data);
+}
+
+static inline void
+tfw_apm_data_get(TfwApmData *data)
+{
+	atomic_inc(&data->refcnt);
+}
+
+static inline void
 __tfw_apm_update(TfwApmRBuf *rbuf, unsigned long jtstamp, unsigned int rtt)
 {
 	int centry = (jtstamp / tfw_apm_jtmintrvl) % rbuf->rbufsz;
 	unsigned long jtmistart = jtstamp - (jtstamp % tfw_apm_jtmintrvl);
 	TfwApmRBEnt *crbent = &rbuf->rbent[centry];
 
 	tfw_apm_rbent_checkreset(crbent, jtmistart);
-	tfw_stats_update(&crbent->pcntrng, rtt, &rbuf->slock);
+	tfw_stats_update(&crbent->pcntrng, rtt);
 }
 
-void
-tfw_apm_update(void *apmdata, unsigned long jtstamp, unsigned long jrtt)
+/*
+ * Calculate the latest percentiles if necessary.
+ * Runs periodically on timer.
+ */
+static void
+tfw_apm_prcntl_tmfn(unsigned long fndata)
 {
-	unsigned int rtt = jiffies_to_msecs(jrtt);
+	int cpu, interval = TFW_APM_TIMER_TIMEOUT;
+	TfwApmData *data, *tmp;
+
+	/* No arguments. */
+	BUG_ON(fndata);
 
-	BUG_ON(!apmdata);
 	/*
-	 * APM stats can't handle response times that are greater than
-	 * the maximum value possible for TfwPcntCtl{}->end. Currently
-	 * the value is USHRT_MAX which is about 65 secs in milliseconds.
+	 * Process work queues on all CPUs and update stats with data
+	 * from each work item in the queue. Add servers with updated
+	 * stats to the list for calculation of stats. Each server is
+	 * is added to the list just once.
+	 *
+	 * If server's APM data is already on the list, that means it
+	 * is on @qrecalc list. Just remove it from @qrecalc list and
+	 * it will be put on @qcalc list as usual for calculation of
+	 * stats values. Note that this is a highly unlikely case.
+	 *
+	 * Note that if server needs a recalculation of stats values,
+	 * it makes sense only if there were updates to server's stats
+	 * data. If there's no updates then a recalculation will lead
+	 * to the same (insufficient) result.
+	 */
+	for_each_online_cpu(cpu) {
+		TfwApmWqItem wq_item;
+		TfwRBQueue *wq = &per_cpu(tfw_apm_wq, cpu);
+
+		while (!tfw_wq_pop(wq, &wq_item)) {
```

Comment:
I am not sure I completely understand you suggestion. Are you suggesting *saving*/*storing* just one RTT sample per millisecond from Tempesta to APM's per-CPU array? Otherwise, an array of just 512 items (which is significantly less than 2k items in WQ now) will not be able to hold all RTT data until the timer function wakes up.

The major point was to avoid concurrency while updating data and calculating stats values for any particular server. That way the data is consistent at the time stats values are calculated. Concurrency caused data inconsistency due to incomplete updates, and data inconsistency caused the stats values calculation failures. So the point was that samples for any particular server should be submitted in a serialized manner (without concurrency), and stats values calculation should be part of that sequence.